### DISCUSSION TITLE
Using MBedCE in Windows 11, over a msys2/ucrt64 shell

### DISCUSSION THREAD

**Discussion by ffxx68** [mbed-ce/mbed-os#390] [Q&A]
I'm a beginner with MBedCE and I'm just setting up my environment, following the hello-world (https://github.com/mbed-ce/mbed-ce-hello-world) and the toolchain (https://github.com/mbed-ce/mbed-os/wiki/Toolchain-Setup-Guide) setups.
I'd like to make use of my current msys2 shell installation in Windows11, as it already includes ninja, python and cmake packages (why install again everything from scratch somewhere else?). 

When going to setup the project from command-line (https://github.com/mbed-ce/mbed-os/wiki/Project-Setup:-Command-Line), and build with:

`$ cmake .. -GNinja -DCMAKE_BUILD_TYPE=Develop -DMBED_TARGET=NUCLEO-LK432KC`

I get this error, which is rather obscure to me:

```
$ cmake .. -GNinja -DCMAKE_BUILD_TYPE=Develop -DMBED_TARGET=NUCLEO-L432KC
-- Mbed: Creating virtual environment with Python interpreter C:/Users/F.Fumi/msys64/usr/bin/python3.12.exe
Error: Refusing to create a venv in C:/Users/F.Fumi/msys64/home/F.Fumi/MBedCE/mbed-ce-hello-world/mbed-os/venv because it contains the PATH separator :.
CMake Error at mbed-os/tools/cmake/mbed_python_interpreter.cmake:37 (execute_process):
  execute_process failed command indexes:

    1: "Child return code: 1"

Call Stack (most recent call first):
  mbed-os/tools/cmake/app.cmake:17 (include)
  CMakeLists.txt:35 (include)

-- Configuring incomplete, errors occurred!
```

This is the shell env PATH:

```
$ env | grep PATH
PKG_CONFIG_PATH=/ucrt64/lib/pkgconfig:/ucrt64/share/pkgconfig
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
MANPATH=/ucrt64/local/man:/ucrt64/share/man:/usr/local/man:/usr/share/man:/usr/man:/share/man
ORIGINAL_PATH=/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/
INFOPATH=/ucrt64/local/info:/ucrt64/share/info:/usr/local/info:/usr/share/info:/usr/info:/share/info
ACLOCAL_PATH=/ucrt64/share/aclocal:/usr/share/aclocal
PKG_CONFIG_SYSTEM_LIBRARY_PATH=/ucrt64/lib
HOMEPATH=\Users\F.Fumi
PATH=/ucrt64/bin:/usr/local/bin:/usr/bin:/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
PKG_CONFIG_SYSTEM_INCLUDE_PATH=/ucrt64/include
```

Any clue, someone?

Thanks
Fabio

**Comment by ffxx68** [comment]
Looks like a consequence of having "C:" in my path. I wonder how to make use of the Unix-like paths.
If only the path had the `C:/Users/F.Fumi/msys64/home/F.Fumi` replaced by `~` in the build scripts... 

How?

**Comment by multiplemonomials** [comment]
Hmm, I tested this on my machine in msys2 mingw64 environment.  I was able to get through venv creation:
```
$ cmake .. -DMBED_TARGET=LPC1768 -DPython3_FIND_REGISTRY=NEVER
-- Building for: Ninja
-- Found Python3: C:/msys64/mingw64/bin/python3.11.exe (found version "3.11.9") found components: Interpreter
-- Mbed: Creating virtual environment with Python interpreter C:/msys64/mingw64/bin/python3.11.exe
-- Found Python3: C:/msys64/home/jamie/dev/mbed-ce-hello-world/mbed-os/venv/bin/python3.11.exe (found version "3.11.9") found components: Interpreter
-- Mbed: Installing Python requirements for Mbed into venv
Requirement already satisfied: pip in c:/msys64/home/jamie/dev/mbed-ce-hello-world/mbed-os/venv/lib/python3.11/site-packages (24.0)
Collecting pip
  Downloading pip-24.3.1-py3-none-any.whl.metadata (3.7 kB)
Downloading pip-24.3.1-py3-none-any.whl (1.8 MB)
   ---------------------------------------- 1.8/1.8 MB 4.5 MB/s eta 0:00:00


<snip>


ERROR: Could not find a version that satisfies the requirement pywin32 (from versions: none)
ERROR: No matching distribution found for pywin32
CMake Error at mbed-os/tools/cmake/mbed_python_interpreter.cmake:58 (execute_process):
  execute_process failed command indexes:

    1: "Child return code: 1"

Call Stack (most recent call first):
  mbed-os/tools/cmake/app.cmake:17 (include)
  CMakeLists.txt:35 (include)


-- Configuring incomplete, errors occurred!

```
(I had to use `-DPython3_FIND_REGISTRY=NEVER` to prevent it from picking up the Windows python instead of the mingw64 python).

I'm not exactly sure what's causing the error you are getting and I am not seeing -- perhaps it's because you are running in the ucrt64 environment while using the MSYS python?  Maybe try installing the ucrt64 python and deleting and recreating your build directory (to clear your CMake cache) and see if that helps at all.

However, from like an hour of trying stuff, I wasn't able to make much progress even without that error happening.  One issue is that the pywin32 package does not seem to exist for the mingw environment, so it has to be commented out from our requirements.txt in order for things to work.  The other issue is that the `cffi` package doesn't build for MinGW due to a bug which is fixed in master branch but not in the latest release.  So you have to clone the `cffi` source code and install it with pip via file path.  However, even then, Pip seems to insist on trying to install the latest source release, so it's impossible to install packages that depend on `cffi`.

Frankly, I think that it will be more than a bit of a project to get this working.  Ultimately the Python packages used by Mbed don't seem to have very good support for msys2 environments, so a lot of hacking like this will be needed to get things running.  I know it's a pain to install stuff manually, but the Python stuff is way better tested if you use the Windows shell, I promise!  Or, if you keep working on this, please let us know how it goes and if you find fixes for these issues!
